src/learner.cc

```diff
@@ -485,10 +485,10 @@ class LearnerImpl : public Learner {
     this->PerformTreeMethodHeuristic(train);
 
     monitor_.Start("PredictRaw");
-    this->PredictRaw(train, &preds_);
+    this->PredictRaw(train, &preds_[train]);
```

Comment:
Given that DMatrix pointers are increasingly used as cache indices, the parameter should probably be changed from `DMatrix*` to `shared_ptr<DMatrix>` in all those places. We can then use `weak_ptr<DMatrix>` as the index into the cache.

This can be done in another pull request, however.